tsdb/chunkenc/chunk.go

```diff
@@ -72,10 +73,22 @@ type Appender interface {
 }
 
 // Iterator is a simple iterator that can only get the next value.
+// Iterator iterates over the samples of a time series.
 type Iterator interface {
+	// Seek advances the iterator forward to the sample with the timestamp t or first value after t.
+	// Seek has no effect if requested timestamp is the same or lower than the current iterator position.
+	// Seek returns false if there is no such sample with the timestamp equal or larger than t.
+	// Iterator is exhausted when the Seek returns false.
+	// TODO(bwplotka): Verify above statements on all implementations with unit test.
+	Seek(t int64) bool
+	// At returns the current timestamp/value pair.
+	// At returns (math.MinInt64, 0.0) before the iterator has advanced.
```

Comment:
So undefined or panic? ((: 

Yes, it just nice to have to have it **consistent**. I vote for this instead of panic, and test in unit test against it. WDYT?

Reply:
I think panic falls under "undefined" :)

If you put that into specification, then all custom iterators need to implement it, because there can be a piece of code relying on it. If there is a good reason for it, that's fine. If not, it's extra burden on implementations.

Reply:
From a quick look, for example `xorIterator` currently doesn't do this, and would require extra code to handle this requirement.

Reply:
This statement is still not true for some iterators in Prometheus (eg. `mergeIterator` will simply panic). I (very strongly) prefer to keep `At()` behaviour before any `Next` or `Seek` operation to be unspecified, so that clients cannot rely on it.

Reply:
I’ll just add... I don’t mind if you change implementations of this interface in Prometheus to be consistent and return (MinInt64, 0.0). (It would be ok to mention that as well, something like “All implementations in Prometheus return (math.MinInt64, 0.0) before first successful Next/Seek, but this isn’t required by external implementations.”

What I object to is adding that requirement to the interface, because then ALL external implementations of this interface (eg. in Thanos or Cortex, and possibly other projects) must do this as well – because it is required by this contract after all. I would even say that it is an incompatible change of API.

Reply:
We're allowed make incompatible changes to APIs, this PR is full of them.

Reply:
That's obviously true, but this change would be very subtle, and doesn't seem to be required to impose on implementors.